[Android] Fix various issues in object dispose methods - Shell edition.

kvpt opened the pull request:
### Description of Change ###
Similar issues to #6467 but for the shell part.

### Issues Resolved ### 
- fixes #6640

### API Changes ###
Removed:
- ShellFragmentContainer => Removed Android Ctor (*)
- HeaderContainer => Removed Android Ctor (*)
- ShellSectionRenderer => Removed Android Ctor (*)
 
_Android Ctor (*)_
`(IntPtr javaReference, JniHandleOwnership transfer) : base(javaReference, transfer)`

### Platforms Affected ### 
- Android

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 
Not applicable

### Testing Procedure ###
Run all of the solution tests cases, there must be no regression.

### PR Checklist ###
- [ ] Has automated tests <!-- (if tests are omitted or manual, state reason in description) -->
- [X] Rebased on top of the target branch at time of PR
- [X] Changes adhere to coding standard

kvpt commented:
It just the first pass.

I seen several others issues with dispose on (ShellItemRenderer, ShellSearchView, ShellToolbarTracker, ShellSectionRenderer).

I will also try to add the reproduction from issue #6640 later.

kvpt commented:
Added the reproduction case.
The UITest doesn't crash after the application go background, I tried to add a wait but it doesn't work.
So the test will be manual only.

kvpt commented:
The second pass is finished, I have removed all suspects using and dispose of ressources being used.

kvpt commented:
Changes can be reviewed. Thanks.

PureWeen commented:
@kvpt can you extract the parts out that apply to https://github.com/xamarin/Xamarin.Forms/issues/6640 ? Or does the whole thing apply to 6640?

kvpt commented:
What cause #6640 is that ShellRenderer dispose (so the parent) is called before ShellSectionRenderer (the child) destroy. And so when the Element is nulled in the ShellRenderer Dispose; after that it crash when ShellSectionRenderer Destroy is called. 

The true fix is the whole thing, explicitly dispose all the things in the right order.

However, if you want a rapid and simple fix, you can just remove this line.
https://github.com/xamarin/Xamarin.Forms/blob/0a157bfd603320b9dac493e85dd16633abc3d4a8/Xamarin.Forms.Platform.Android/Renderers/ShellRenderer.cs#L361

It will not correct the true problem but will fix the crash.
It's up to you if this alternative solution is acceptable or not before this pull request get merged.

PureWeen commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue6640.cs:
```diff
@@ -0,0 +1,95 @@
+{
+	[Preserve(AllMembers = true)]
+	[Issue(IssueTracker.Github, 6640, "[Android] Crash when app go background", PlatformAffected.Android)]
+	public class Issue6640 : TestContentPage
```
Can you add a UITest for this?

kvpt replied:
See my comment : https://github.com/xamarin/Xamarin.Forms/pull/6715#issuecomment-507080479

PureWeen replied:
Now that unhook is part of dispose (https://github.com/xamarin/Xamarin.Forms/pull/6715/files#diff-edc32d3b021f02879f303175242f36b7R174) instead of DestroyView is there a way to emulate this behavior without the backgrounding? Perhaps after the mainpage is swapped out https://github.com/xamarin/Xamarin.Forms/pull/6715/files#diff-fbd24038d82294fdd7a696c78d6e02d4R59 the item could be removed from the collection to trigger the dispose? 

```
Application.Current.MainPage = new ContentPage
														{
															Content = new StackLayout
															{
																BackgroundColor = Color.White,
																Children =
																{
																	new Label
																	{
																		Text = $"Press Back Arrow to send application to background and wait few seconds.{Environment.NewLine}Application must not crash.",
																		AutomationId = "ReadyToGoBack"
																	}
																}
															}
														};

														Device.BeginInvokeOnMainThread(() =>
														{
															this.CreateContentPage();
															this.Items.RemoveAt(0);
														});
```

or something similar to that? We just need to trigger the destroy or dispose in order to have a UI test for this right? because that's all the backgrounding is really doing

kvpt replied:
I think the only way to trigger the bug is on Activity Destroy.
Because it involve two methods :
- ShellRenderer.Dispose : Called when the main page change for exemple
- ShellSectionRenderer.OnDestroy : Called .... when Android want, and from this [stackoverflow answer](https://stackoverflow.com/a/17195813) it's only garanted when the fragment activity is also destroyed. This statement is coherent with the issue reproduction which is only triggered on activity destroy.
It's for that reason that I have replaced all the OnDestroy with Dispose.

PureWeen commented on Xamarin.Forms.Platform.Android/Renderers/ShellBottomNavViewAppearanceTracker.cs:
```diff
@@ -135,20 +131,20 @@ public void Dispose()
-				{
-					_defaultList?.Dispose();
-				}
+				_disposed = true;
```
Can we move all the _disposed checks outside of this if loop like we try to do most other places

https://github.com/xamarin/Xamarin.Forms/blob/4187c67970480340d504111520462d8bb42f7a85/Xamarin.Forms.Platform.Android/AppCompat/MasterDetailPageRenderer.cs#L210

kvpt replied:
It's done.

kvpt commented:
Fixed merge conflicts and rebased on 4.2.0.

kvpt commented:
Rebased again on 4.2.0 to fix conflicts.

Also fixed a new issue introduced in #6572.
https://github.com/xamarin/Xamarin.Forms/blob/069f68e8b74ead5cfa5356ed6e75c7af54db34e4/Xamarin.Forms.Platform.Android/Renderers/ShellFlyoutTemplatedContentRenderer.cs#L257

kvpt commented:
I fixed a regression which caused an NRE during the dispose of the ShellToolbarTracker (Issue4684).
I runned locally all the tests and all the tests related to the shell are ok.

alanag13 commented:
any idea when I can expect this to be merged? I'm running into some of the issues that this branch fixes. Thanks for your help @kvpt !

kvpt commented:
Fixed conflicts and rebased on master (since it's the new branch targeted).

dlakes-hx commented:
Is the intent that this will be part of a 4.2 service release, or part of 4.3?

rmarinho commented:
This one fixes #7327

rmarinho commented:
@kvp we want to pull this in for 4.2.0 service release, do you mind rebase it again on top of 4.2.0 ? 

thanks sorry going back and forth

kvpt commented on Xamarin.Forms.Platform.Android/Renderers/ShellSearchView.cs:
```diff
@@ -123,87 +124,98 @@ bool TextView.IOnEditorActionListener.OnEditorAction(TextView v, ImeAction actio
-			_clearPlaceholderButton = null;
-			_shellContext = null;
-			_searchHandlerAppearanceTracker = null;
+				_textBlock.Adapter = null;
```
Good catch, fixed.

kvpt commented:
@rmarinho No problem, I rebased my changes onto 4.2.0 branch and applied the remarks.
But I doesn't have the right to switch the target branch of the PR so you will have to do it.

PureWeen commented:
Changed base and kicked off UI Tests

PureWeen commented on Xamarin.Forms.Platform.Android/Renderers/ShellFlyoutTemplatedContentRenderer.cs:
```diff
@@ -191,10 +194,7 @@ protected virtual void UpdateFlyoutHeaderBehavior()
-
-			if (_flyoutHeader != null)
-				margin = _flyoutHeader.Margin;
+			var margin = _flyoutHeader?.Margin ?? default;
```
@kvpt we don't have  C# 7.3 on our 4.2 branch it's only 7.0 so can you just change this to default(Thickness) ?

PureWeen replied:
```suggestion
			var margin = _flyoutHeader?.Margin ?? default(Thickness);
```

PureWeen commented:
@samhouts I think out of that list the main ones to fix would be

> public event System.EventHandler Destroyed;

Keep this

> protected virtual void OnItemsCollectionChagned

I'm wondering if we should just obsolete this one given that it's a spelling error 👎 and then keep the new one



These constructors

> protected ShellSectionRenderer (IntPtr javaReference, Android.Runtime.JniHandleOwnership transfer);

are only used if there's a memory leak so I think it's better to get rid of them

> public override void OnDestroy ();

these are just overrides so they should be fine

kvpt commented:
I did it on purpose, I had to add the IDisposable interface to IShellFlyoutRenderer which is a breaking change, so I don't tried to limit the breaking changes after that because they are all compilation breaking changes.

> IShellFlyoutRenderer => Inherit IDisposable

I don't think this one can be avoided

> ShellSectionRenderer.OnItemsCollectionChagned =>ShellSectionRenderer.OnItemsCollectionChanged

I can obsolete it and add the correct one, but as there is one breaking change it's an opportunity to remove it now, faster this code is removed fewer libraries and user code are impacted

> ShellFragmentContainer - HeaderContainer - ShellSectionRenderer => Removed Android Ctor

These constructors are poisons, they add dispose problem, when Android try to restore an object with this constructor. Because when you try to acces an object disposed on the java side, instead of crashing, it silently continue, and create a memory leak.

> IShellItemRenderer - ShellItemRendererBase => Removed Destroyed event

At first I did not remove it, but after some thought, I decided to remove it, because now that we are not linked to destroy event it make no sense at all to have it. But we can obsolete them if necessary, for this one it's logic to do that if we want to be more conservative on breaking the contract.

PureWeen commented:
API 19 UI Test failure: FlyoutNavigationBetweenItemsWithNavigationStacks_lg_nexus_5

> I did it on purpose, I had to add the IDisposable interface to IShellFlyoutRenderer which is a breaking change, so I don't tried to limit the breaking changes after that because they are all compilation breaking changes.

Do you have to add IDisposable to IShellFlyoutRenderer? The Android implementation of IShellFlyoutRenderer already implements IDisposable so can we cast IShellFlyoutRenderer to *IDisposable* and if it's not null then call dispose.

https://github.com/xamarin/Xamarin.Forms/pull/6715/files#diff-ea33147016e2d6ca6e3029394a4cb3f7R361


> I don't think this one can be avoided
> ShellSectionRenderer.OnItemsCollectionChagned =>ShellSectionRenderer.OnItemsCollectionChanged
> I can obsolete it and add the correct one, but as there is one breaking change it's an opportunity to remove it now, faster this code is removed fewer libraries and user code are impacted

I would still just obsolete it for now and we can look at removing it when we bump up to 5.0

> ShellFragmentContainer - HeaderContainer - ShellSectionRenderer => Removed Android Ctor

> These constructors are poisons, they add dispose problem, when Android try to restore an object with this constructor. Because when you try to acces an object disposed on the java side, instead of crashing, it silently continue, and create a memory leak.

These are fine to remove


> IShellItemRenderer - ShellItemRendererBase => Removed Destroyed event
> At first I did not remove it, but after some thought, I decided to remove it, because now that we are not linked to destroy event it make no sense at all to have it. But we can obsolete them if necessary, for this one it's logic to do that if we want to be more conservative on breaking the contract.

Obsolete them. 

96.3 percent of the time that we break the ABI we have to go back and fix it anyway so it's better to just be cautious about it now

kvpt commented:
Ok I will try for the IDisposable change, and obsolete the others changes.

kvpt commented:
I reverted most breaking changes, the only remaining are the specific android contructors.

@samhouts For the 5.0 please consider to add the IDisposable to the interface IShellFlyoutRenderer, iOS version have it.

@PureWeen For the failed test, I executed the test 6738 on an api 19 emulator and it works fine for me. However I executed it manually because the test don't want to start on my machine currently.
Manually deploy the test app worked before to enable the tests to run, but not now, strange.

kvpt commented:
Ok from the stacktrace, it's clear it's a problem with dispose.
I think it's because ColorChangeRevealDrawable need to be disposed.

https://github.com/xamarin/Xamarin.Forms/blob/4fd2d6b45b7f287433ea2bf52581d376a06b999f/Xamarin.Forms.Platform.Android/Renderers/ColorChangeRevealDrawable.cs#L26

This part in the constructor launch an async operation, and if we dispose the BottomNavigationView before the operation finish, I think it can cause this issue. 

I will test that.

kvpt commented:
@PureWeen Can you relaunch UI tests please
I implemented the dispose of the ColorChangeRevealDrawable class to force the animation to end before disposing. I think it's what cause the race condition. The tests will confirm that or not.

PureWeen commented on Xamarin.Forms.Platform.Android/Renderers/ShellBottomNavViewAppearanceTracker.cs:
```diff
@@ -80,7 +81,9 @@ protected virtual void SetBackgroundColor(BottomNavigationView bottomView, Color
-				bottomView.SetBackground(new ColorChangeRevealDrawable(_lastColor.ToAndroid(), color.ToAndroid(), touchPoint));
+				_colorChangeRevealDrawable = new ColorChangeRevealDrawable(_lastColor.ToAndroid(), color.ToAndroid(), touchPoint);
+
+				bottomView.SetBackground(_colorChangeRevealDrawable);
```
maybe this should be set to null on dispose?

kvpt replied:
I changed my approach.
I now only dispose the old background after it has been changed.
The final dispose is done in ShellItemRenderer.
I succeed to reproduce when disposing the AppearanceTracker before the bottom view background. But in the current order it work fine. So not sure if it will work.

kvpt commented:
@PureWeen Can you restart another batch of tests please.

If this time the exception occur again I will have the Xamarin Android "Revive" constructor and throw an exception with the callstack in it to see where the access on the disposed background is done.

PureWeen commented:
Alright @kvpt let's see if this works

The part of the UI test that's failing performs two operations
- inserts a page in the back stack
- then switches tabs

Both of those cause a SendAppearenceChanged event to fire so both of those cause the *ColorChangeRevealDrawable* to get created then destroyed. 

I'm guessing there's a race condition where the first one is being drawn after the second *SendAppearenceChanged* disposes. 

I changed the code so it doesn't try to set the background again if the color hasn't changed. It might even make sense to not call *SendAppearenceChanged* when InsertPageBefore is called as well

kvpt commented:
@PureWeen It make sense, this kind of issue is always a race condition.
And yes I also think it's useless to call AppearanceChanged if the page has not been rendered yet.

PureWeen commented:
Dang that still didn't work 👎

PureWeen commented:
@kvpt so on the last round I commented out the dispose completely
https://github.com/xamarin/Xamarin.Forms/pull/6715/commits/1e135afd157de4da9c6bd5c9130e6442e60da4b9#diff-0b13b4b03eb6fc03312913a8efabf296R94

And still the same exception is popping up

kvpt commented:
If you want to remove the dispose, the problematic one is here : 
https://github.com/kvpt/Xamarin.Forms/blob/1e135afd157de4da9c6bd5c9130e6442e60da4b9/Xamarin.Forms.Platform.Android/Renderers/ShellItemRenderer.cs#L100

And now i see this again, I think it's useless, because the dispose of the BottowView will surely dispose the background it self.

PureWeen commented:
@kvpt  💚 💚💚💚💚💚💚💚

Your last suggestion looks like it worked!

https://github.com/xamarin/Xamarin.Forms/pull/6715/commits/e73a0608b2b6809610f119e55cb0b5cbd2b17477#diff-4144388f1cafc65a54c1026775d46ab3R99

I'll refine and test a bit next week and then we should be able to get this merged

PureWeen commented:
Alright all the UI Tests relevant to this PR passed. I rebased with the latest 4.2.0 code and kicked them off for good measure but this PR should be good to review now.

rmarinho commented on Xamarin.Forms.Platform.Android/Renderers/IShellItemRenderer.cs:
```diff
@@ -10,6 +9,7 @@ public interface IShellItemRenderer : IDisposable
 
 		ShellItem ShellItem { get; set; }
 
+		[Obsolete("Destroyed event is obsolete as of version 4.2.0.")]
```
We need to add this to release notes for sr4

rmarinho commented on Xamarin.Forms.Platform.Android/Renderers/ShellBottomNavViewAppearanceTracker.cs:
```diff
@@ -135,20 +138,23 @@ public void Dispose()
 		}
 
-#endregion IDisposable
+		#endregion IDisposable
```
```suggestion
#endregion IDisposable
```

rmarinho commented on Xamarin.Forms.Platform.Android/Renderers/ShellItemRenderer.cs:
```diff
@@ -253,10 +255,18 @@ protected virtual void OnMoreItemSelected(ShellSection shellSection, BottomSheet
+			if (_bottomSheetDialog != null)
+			{
+				_bottomSheetDialog.DismissEvent -= OnMoreSheetDismissed;
+				_bottomSheetDialog.Dispose();
```
```suggestion
				_bottomSheetDialog.Dispose();
				_bottomSheetDialog = null;
```

rmarinho commented on Xamarin.Forms.Platform.Android/Renderers/ShellItemRendererBase.cs:
```diff
@@ -23,14 +22,16 @@ ShellItem IShellItemRenderer.ShellItem
-		public event EventHandler Destroyed;
-
+		[Obsolete("Destroyed event is obsolete as of version 4.2.0.")]
+		public event EventHandler Destroyed;	
```
Needs to be added to release notes

rmarinho commented on Xamarin.Forms.Platform.Android/Renderers/ShellRenderer.cs:
```diff
@@ -332,15 +319,14 @@ public override void Draw(Canvas canvas)
+					paint.Color = _color;
 
-				canvas.DrawRect(new Rect(0, bounds.Bottom - _bottomSize, bounds.Right, bounds.Bottom), paint);
+					canvas.DrawRect(new Rect(0, 0, bounds.Right, _topSize), paint);
```
why two DrawRect ?

rmarinho commented on Xamarin.Forms.Platform.Android/Renderers/ShellRenderer.cs:
```diff
@@ -361,19 +347,24 @@ void IDisposable.Dispose()
+				(_flyoutRenderer as IDisposable)?.Dispose();
 
-				_disposed = true;
+				_currentRenderer.Dispose();
```
```suggestion
				_currentRenderer.Dispose();
				_currentRenderer = null;
```

rmarinho commented on Xamarin.Forms.Platform.Android/Renderers/ShellToolbarTracker.cs:
```diff
@@ -453,34 +454,39 @@ protected virtual void UpdateToolbarItems(Toolbar toolbar, Page page)
-			foreach (var item in page.ToolbarItems)
+			if (page != null)
 			{
-				using (var title = new Java.Lang.String(item.Text))
```
Why did we changed this from using? any special issue?

PureWeen commented:
@kvpt Thank you for all the work on this one!!

We've merged in the slightly slimmed down version of this here
https://github.com/xamarin/Xamarin.Forms/pull/7768

I'm going to close this for now. 

If there are still aspects of this PR you think are important to apply then can you rebase and we'll reopen?

kvpt commented:
@PureWeen Thank you too.
Don't have much time to follow this at the moment.
I replaced destroy with dispose call because we can't control when android call destroy, but the root cause of the problem was the dispose/destroy order which was not right.
If this part is ok now, it's ok for me.
I also see that you kept the "leak constructor" removal, that was an important thing for me.
Thank you again.
